ksql-api/src/main/java/io/confluent/ksql/api/server/ReactiveSubscriber.java

```diff
@@ -0,0 +1,176 @@
+
+  protected final Context context;
+  private Subscription subscription;
+  private boolean complete;
+
+  /**
+   * Construct a ReactiveSubscriber
+   *
+   * @param context The Vert.x context to use for the subscriber - the subscriber code will always
+   *                be executed on this context. This ensures the code is never executed
+   *                concurrently by more than one thread.
+   */
+  public ReactiveSubscriber(final Context context) {
+    this.context = context;
+  }
+
+  @Override
+  public void onSubscribe(final Subscription s) {
+    Objects.requireNonNull(s);
+    context.runOnContext(v -> doOnSubscribe(s));
+  }
+
+  @Override
+  public void onNext(final T t) {
+    Objects.requireNonNull(t);
+    context.runOnContext(v -> doOnNext(t));
+  }
+
+  @Override
+  public void onError(final Throwable t) {
+    Objects.requireNonNull(t);
+    context.runOnContext(v -> doOnError(t));
+  }
+
+  @Override
+  public void onComplete() {
+    context.runOnContext(v -> doOnComplete());
+  }
+
+  protected void afterSubscribe(final Subscription subscription) {
+  }
+
+  protected void handleValue(final T t) {
+  }
+
+  protected void handleComplete() {
+  }
+
+  protected void handleError(final Throwable t) {
+  }
+
+  private void doOnSubscribe(final Subscription subscription) {
+    checkContext();
+    if (this.subscription != null) {
+      try {
+        // Cancel the new subscription
+        subscription.cancel();
+      } catch (final Throwable t) {
+        final Exception e =
+            new IllegalStateException("Exceptions must not be thrown from cancel");
+        logError(e);
+      }
+    }
+    this.subscription = subscription;
+    afterSubscribe(subscription);
+  }
+
+  private void doOnNext(final T val) {
+    checkContext();
+    if (complete) {
+      return;
+    }
+    if (subscription == null) {
+      final Exception e = new IllegalStateException(
+          "onNext must be called without request being called");
+      logError(e);
+    }
+    try {
+      handleValue(val);
+    } catch (final Throwable t) {
+      complete();
+      onError(t);
+    }
+  }
+
+  private void doOnError(final Throwable t) {
+    checkContext();
+    if (subscription == null) {
+      logError(new IllegalStateException("onError must not be called before onSubscribe"));
+    } else {
+      complete = true;
+      handleError(t);
+    }
+  }
+
+  private void doOnComplete() {
+    checkContext();
+    if (subscription == null) {
+      logError(new IllegalStateException("onComplete must not be called before onSubscribe"));
+    } else {
+      complete = true;
+      handleComplete();
+    }
+  }
+
+  protected void makeRequest(final long l) {
+    checkContext();
+    try {
+      subscription.request(l);
+    } catch (Throwable t) {
+      final Exception e =
+          new IllegalStateException("Exceptions must not be thrown from request");
+      logError(e);
+    }
+  }
+
+  protected void complete() {
```

Comment:
I tend to err on the side of keeping the method non final unless there is a very good reason not to. There can be valid reasons for overriding that we haven't considered and we don't want to reduce the value of this class to subclass implementers

Reply:
I think there are some very good reasons 😉 this method, for example, changes private state that this class makes certain assumptions about - if the behavior of `complete()` was any different than what is written in this class, then other methods would break (e.g. `doOnNext` would keep handling values). In general I think keeping the surface area of the API as small as possible at the start and expanding it only when necessary is safer.

Reply:
Made them final